rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -1278,4 +1701,230 @@ mod tests {
+                .col("key", array::rand_pseduo_uuid_hex());
+            let batch = data.into_batch_rows(RowCount::from(1024)).unwrap();
+            let batch1 = batch.slice(0, 512);
+            let batch2 = batch.slice(512, 512);
+            let schema = batch.schema();
+
+            let reader = Box::new(RecordBatchIterator::new(
+                [Ok(batch1.clone())],
+                schema.clone(),
+            ));
+            let write_params = WriteParams {
+                max_rows_per_file: 256,
+                max_rows_per_group: 32, // Non-standard group size to hit edge cases
+                ..Default::default()
+            };
+            let mut ds = Dataset::write(reader, "memory://", Some(write_params.clone()))
+                .await
+                .unwrap();
+
+            if scalar_index {
+                let index_params = ScalarIndexParams::default();
+                ds.create_index(&["key"], IndexType::Scalar, None, &index_params, false)
+                    .await
+                    .unwrap();
+            }
+
+            // Another two batches, not in the scalar index (if there is one)
+            let reader = Box::new(RecordBatchIterator::new(
+                [Ok(batch2.clone())],
+                batch2.schema(),
+            ));
+            ds.append(reader, Some(write_params)).await.unwrap();
+
+            let ds = Arc::new(ds);
+
+            // New data with only a subset of columns
+            let update_schema = Arc::new(schema.project(&[2, 1]).unwrap());
+            // Full second file and part of third file.
+            let indices: Int64Array = (256..512).chain(600..612).chain([712, 715]).collect();
+            let keys = arrow::compute::take(batch["key"].as_ref(), &indices, None).unwrap();
+            let new_data = RecordBatch::try_new(
+                update_schema.clone(),
+                vec![
+                    keys,
+                    Arc::new((1000..(1000 + indices.len() as u32)).collect::<UInt32Array>()),
+                ],
+            )
+            .unwrap();
+
+            Fixtures { ds, new_data }
+        }
+
+        #[tokio::test]
+        async fn test_delete_not_supported() {
+            let Fixtures { ds, new_data } = setup(false).await;
+
+            let reader = Box::new(RecordBatchIterator::new(
+                [Ok(new_data.clone())],
+                new_data.schema(),
+            ));
+
+            // Should reject when_not_matched_by_source_delete as not yet supported
+            let job = MergeInsertBuilder::try_new(ds.clone(), vec!["key".to_string()])
+                .unwrap()
+                .when_not_matched_by_source(WhenNotMatchedBySource::Delete)
+                .when_matched(WhenMatched::UpdateAll)
+                .when_not_matched(WhenNotMatched::DoNothing)
+                .try_build()
+                .unwrap();
+            let res = require_send(job.execute_reader(reader)).await;
```

Comment:
Is the `require_send` strictly needed or was it for debugging purposes?

Or is it a form of assert?

Reply:
Yeah meant to be a form of assertion. Our Python bindings require the async functions we expose to be Send so we can run them in the executor. I'll rename `assert_send()`.